src/v/utils/xid.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+#include "bytes/iobuf_parser.h"
+
+#include <seastar/core/sstring.hh>
+
+#include <fmt/core.h>
+
+#include <array>
+#include <cstdint>
+#include <exception>
+
+/**
+ * Exception indicating that provided XID isn't valid f.e. incorrect length of
+ * invalid characters
+ */
+class invalid_xid final : public std::exception {
+public:
+    explicit invalid_xid(const ss::sstring&);
+    const char* what() const noexcept final { return _msg.c_str(); }
+
+private:
+    ss::sstring _msg;
```

Comment:
if the code points are static, i.e.: `throw invalid_xid("foo-bar-baz")` then you could simply take a string_view 

maybe a `invalid_xid_view` that takes in a stringview rather than a copy w/ allocation.

Reply:
in this case the message is formatted when instantiating the exception object.